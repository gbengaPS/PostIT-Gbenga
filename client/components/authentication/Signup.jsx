import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import validateSignUpInput from '../../../server/shared/validateSignUpInput';

import { signupLoading, signupUser } from '../../actions/userAuthActions';
import SignupForm from './SignupForm';

/**
 *
 * @class Signup
 *
 * @extends { React.Component }
 */
class Signup extends React.Component {

  /**
   * @description Creates an instance of Signup
   *
   * @param { object } props -react props object
   *
   * @returns { void } -return void
   */
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.state = { errors: {} };
  }
  /**
   * @param { object } event -event object
   *
   * @returns { void } -returns nothing
   */
  handleChange(event) {
    const value = event.target.value.trim();
    const name = event.target.name;
    this.setState({ [name]: value });
  }
   /**
   * @param { object } event -event objet
   *
   * @returns { void } -returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    this.setState({ errors: {} });
    const { errors, isValid } = validateSignUpInput(this.state);
    if (!isValid) {
      this.setState({ errors });
    } else {
      this.props.setLoading(true);
      this.props.signupUser(this.state, this.props.history);
    }
  }
  /**
   *
   * @returns { void } -returns react element
   */
  render() {
    return (
      <SignupForm
        loading={this.props.isLoading}
        error={this.props.error}
        errors={this.state.errors}
        validate={this.validateInput}
        handleSubmit={this.handleSubmit}
        handleChange={this.handleChange}
      />
    );
  }
}
const mapStateToProps = state => (
  {
    user: state.authReducer.user,
    error: state.authReducer.signupError,
    isLoading: state.itemLoadingReducer.signupLoading,
  }
);
const mapDispatchToProps = dispatch => (
  {
    signupUser: (user, history) => {
      dispatch(signupUser(user, history));
    },
    setLoading: (bool) => {
      dispatch(signupLoading(bool));
    },
  }
);
Signup.propTypes = {
  user: PropTypes.object,
  isLoading: PropTypes.bool,
  setLoading: PropTypes.func.isRequired,
  signupUser: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  error: PropTypes.string,
};
export default connect(mapStateToProps, mapDispatchToProps)(Signup);
