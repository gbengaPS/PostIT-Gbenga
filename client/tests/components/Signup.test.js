import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import 'mock-local-storage';
import Signup from '../../components/authentication/Signup';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Signup component', () => {
  const store = mockStore({
    user: {},
    authReducer: {},
    itemLoadingReducer: {}
  });
  const props = {
    history: {}
  };
  const SignupWrapper = shallow(<Signup store={store} {...props}/>);
  it('should render correctly', () => {
    expect(SignupWrapper).toMatchSnapshot();
  });
  // it('should handle change event', () => {
  //   const mockEvent = {};
  //   console.log(SignupWrapper.instance());
  //   expect(SignupWrapper.dive().instance().handleClick(mockEvent)).toBe(true);
  // });
  it('should not submit when for contains errors', () => {
    
  });
});
