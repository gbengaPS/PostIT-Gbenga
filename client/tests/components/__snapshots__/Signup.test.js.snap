// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Signup
    history={Object {}}
    setLoading={[Function]}
    signupUser={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "error": undefined,
      "errors": Object {},
      "handleChange": [Function],
      "handleSubmit": [Function],
      "loading": undefined,
      "validate": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "error": undefined,
        "errors": Object {},
        "handleChange": [Function],
        "handleSubmit": [Function],
        "loading": undefined,
        "validate": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Signup component should render correctly when connected to the store 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Signup)
    history={Object {}}
    setLoading={[Function]}
    signupUser={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "error": undefined,
      "history": Object {},
      "isLoading": undefined,
      "setLoading": [Function],
      "signupUser": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "user": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "error": undefined,
        "history": Object {},
        "isLoading": undefined,
        "setLoading": [Function],
        "signupUser": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "user": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
